router-tests/testenv/testexec.go

```diff
@@ -0,0 +1,136 @@
+package testenv
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"github.com/hashicorp/consul/sdk/freeport"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/atomic"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+)
+
+const routerDir = "../router"
+
+// RunRouterBinary starts the router binary, sets up the test environment, and runs the provided test function.
+func RunRouterBinary(t *testing.T, cfg *Config, f func(t *testing.T, xEnv *Environment)) {
+	t.Helper()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	buildRouterBin(t, ctx)
+	env := runRouterBin(t, ctx, cfg, routerDir)
+	t.Cleanup(env.Shutdown)
+
+	// Execute the test case with the environment
+	f(t, env)
+}
+
+// BuildRouter runs `make build` inside the router directory and fails the test on error.
+func buildRouterBin(t *testing.T, ctx context.Context) {
+	t.Helper()
```

Comment:
How do we avoid building the binary multiple times as soon as we add more tests? I recommend to build the binary once even when you call it multiple times. Additionally, we should build it with the right build flags including race.